crates/sui-json-rpc/src/balance_changes.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::BTreeMap;
+use std::ops::Neg;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use move_core_types::language_storage::TypeTag;
+use tokio::sync::RwLock;
+
+use sui_core::authority::AuthorityState;
+use sui_json_rpc_types::{BalanceChange, BalanceChangeType};
+use sui_types::base_types::{MoveObjectType, ObjectID, ObjectRef, SequenceNumber};
+use sui_types::coin::Coin;
+use sui_types::error::SuiError;
+use sui_types::gas::GasCostSummary;
+use sui_types::gas_coin::GAS;
+use sui_types::messages::TransactionEffectsAPI;
+use sui_types::messages::{ExecutionStatus, TransactionEffects};
+use sui_types::object::{Object, Owner};
+use sui_types::storage::WriteKind;
+
+pub async fn get_balance_change_from_effect<P: ObjectProvider<Error = E>, E>(
+    object_provider: &P,
+    effects: &TransactionEffects,
+) -> Result<Vec<BalanceChange>, E> {
+    let (_, gas_owner) = effects.gas_object();
+
+    // Only charge gas when tx fails, skip all object parsing
+    if effects.status() != &ExecutionStatus::Success {
+        return Ok(vec![BalanceChange {
+            owner: *gas_owner,
+            change_type: BalanceChangeType::Gas,
+            coin_type: GAS::type_tag(),
+            amount: effects.gas_cost_summary().net_gas_usage().neg() as i128,
+        }]);
+    }
+
+    let all_mutated: Vec<(&ObjectRef, &Owner, WriteKind)> = effects.all_mutated();
+    let all_mutated = all_mutated
+        .iter()
+        .map(|((id, version, _), _, _)| (*id, *version))
+        .collect::<Vec<_>>();
+
+    get_balance_change(
+        object_provider,
+        *gas_owner,
```

Comment:
So we only return the balance change for the transaction sender, but not anyone else?

Reply:
we do, we will have 1 balance change per address, sender and recipients.